tools/clang/include/clang/SPIRV/SpirvVisitor.h

```diff
@@ -0,0 +1,182 @@
+  virtual bool visit(SpirvStore *) {}
+  virtual bool visit(SpirvUnaryOp *) {}
+  virtual bool visit(SpirvVectorShuffle *) {}
+
+protected:
+  Visitor() = default;
+};
+
+/// \breif The visitor class that emits the SPIR-V words from the in-memory
+/// representation.
+class EmitVisitor : public Visitor {
+public:
+  EmitVisitor() : Visitor(), spirvBinary({}) {}
+
+  // Visit different SPIR-V constructs for emitting.
+  bool visit(SpirvModule *, Phase phase);
+  bool visit(SpirvFunction *, Phase phase);
+  bool visit(SpirvBasicBlock *, Phase phase);
+
+  bool visit(SpirvCapability *);
+  bool visit(SpirvExtension *);
+  bool visit(SpirvExtInstImport *);
+  bool visit(SpirvMemoryModel *);
+  bool visit(SpirvEntryPoint *);
+  bool visit(SpirvExecutionMode *);
+  bool visit(SpirvString *);
+  bool visit(SpirvSource *);
+  bool visit(SpirvName *);
+  bool visit(SpirvModuleProcessed *);
+  bool visit(SpirvDecoration *);
+  bool visit(SpirvVariable *);
+  bool visit(SpirvFunctionParameter *);
+  bool visit(SpirvLoopMerge *);
+  bool visit(SpirvSelectionMerge *);
+  bool visit(SpirvBranching *);
+  bool visit(SpirvBranch *);
+  bool visit(SpirvBranchConditional *);
+  bool visit(SpirvKill *);
+  bool visit(SpirvReturn *);
+  bool visit(SpirvSwitch *);
+  bool visit(SpirvUnreachable *);
+  bool visit(SpirvAccessChain *);
+  bool visit(SpirvAtomic *);
+  bool visit(SpirvBarrier *);
+  bool visit(SpirvBinaryOp *);
+  bool visit(SpirvBitFieldExtract *);
+  bool visit(SpirvBitFieldInsert *);
+  bool visit(SpirvComposite *);
+  bool visit(SpirvCompositeExtract *);
+  bool visit(SpirvExtInst *);
+  bool visit(SpirvFunctionCall *);
+  bool visit(SpirvNonUniformBinaryOp *);
+  bool visit(SpirvNonUniformElect *);
+  bool visit(SpirvNonUniformUnaryOp *);
+  bool visit(SpirvImageOp *);
+  bool visit(SpirvImageQuery *);
+  bool visit(SpirvImageSparseTexelsResident *);
+  bool visit(SpirvImageTexelPointer *);
+  bool visit(SpirvLoad *);
+  bool visit(SpirvSampledImage *);
+  bool visit(SpirvSelect *);
+  bool visit(SpirvSpecConstantBinaryOp *);
+  bool visit(SpirvSpecConstantUnaryOp *);
+  bool visit(SpirvStore *);
+  bool visit(SpirvUnaryOp *);
+  bool visit(SpirvVectorShuffle *);
+
+private:
+  // Initiates the creation of a new instruction with the given Opcode.
+  // Providing a size hint can slightly improve performance for instructions
+  // that have fixed size by reserving the needed vector size.
+  void initInstruction(spv::Op, uint32_t sizeHint = 0);
+
+  // Finalizes the current instruction by encoding the instruction size into the
+  // first word, and then appends the current instruction to the SPIR-V binary.
+  void finalizeInstruction();
+
+  // Encodes the given string into the current instruction that is being built.
+  void encodeString(llvm::StringRef value);
+
+  // Provides the next available <result-id>
+  uint32_t getNextId() { return ++id; }
+
+private:
+  uint32_t id;
+  std::vector<uint32_t> curInst;
```

Comment:
I think we can use a `SmallVector<uint32_t, 16>` here. We are unlikely to have instructions very large (except `OpSource` & `OpSourceContinued` maybe). We can put them in stack entirely. And we can also drop the hint thing by doing this.